jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,212 @@
+  static Optional<String> performUpdateCheck(
+      Consumer<LogEvent> log, String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    Path configFile = configDir.resolve("config.json");
+    Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+
+    try {
+      // Check global config
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        try {
+          ConfigJsonTemplate config =
+              JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+          if (config.disableUpdateCheck) {
+            return Optional.empty();
+          }
+        } catch (IOException ex) {
+          log.accept(
+              LogEvent.warn(
+                  "Failed to read global Jib config: "
+                      + ex.getMessage()
+                      + "; you may need to fix or delete "
+                      + configFile
+                      + "; "));
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        try (OutputStream outputStream = Files.newOutputStream(configFile)) {
+          JsonTemplateMapper.writeTo(config, outputStream);
+        } catch (IOException ex) {
+          // If attempt to generate new config file failed, delete so we can try again next time
+          log.accept(LogEvent.debug("Failed to generate global Jib config; " + ex.getMessage()));
+          Files.deleteIfExists(configFile);
+        }
+      }
+
+      // Check time of last update check
+      if (Files.exists(lastUpdateCheck)) {
+        try {
+          String fileContents =
+              new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException | IOException ex) {
+          // If reading update time failed, file might be corrupt, so delete it
+          log.accept(LogEvent.debug("Failed to read lastUpdateCheck; " + ex.getMessage()));
+          Files.delete(lastUpdateCheck);
+        }
+      }
+
+      // Check for update
+      HttpURLConnection connection = (HttpURLConnection) new URL(versionUrl).openConnection();
+      try {
+        connection.setConnectTimeout(3000);
+        BufferedReader bufferedReader =
+            new BufferedReader(
+                new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+        VersionJsonTemplate version =
+            JsonTemplateMapper.readJson(
+                bufferedReader.readLine().trim(), VersionJsonTemplate.class);
```

Comment:
`readLine()` seems unsafe. It will break if our JSON file is not a single-line JSON. I think you used `BufferedReader` only to use `readLine()`, so perhaps it's no longer necessary. Maybe use `InputStream` directly?